enhancements/olm-deletes-operands.md

```diff
@@ -0,0 +1,155 @@
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: none
+```
+
+##### OLM managed
+
+With `operandCleanup.type: olm-managed` OLM will delete all CRs, for the operator's owned APIs, in the target namespaces (`olm.targetNamespaces`) of that operator.
+Once the CRs have their finalizers cleared by the operator and are removed from the foreground, OLM will delete the CRD and then proceed with the removal of the operator.
+
+```YAML
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+spec:
+  ...
+  operandCleanup:
+    type: olm-managed
+    timeout: 30s
+```
+
+Additionaly a timeout configured by `spec.operandCleanup.timeout` would specify how long OLM should wait for the operator to finish cleaning up.
+
+If the cleanup timeout is exceeded, OLM should generate an alert event indicating a cleanup failure for that operator, and then proceed to uninstall the operator.
+
+**Note:** See [Risks and Mitigations section][risks-section] on configuring the timeout.
+
+#### Cleanup for namespace-scoped operators
+
+If there are two non-global operators installed of the same type (providing the same APIs), then removing the CRD during the cleanup of one operator would result in the deletion of CRs for the other operator as well.
+
+To avoid this case OLM would have to ensure that it only removes the CRD during cleanup if:
+
+- The operator installed is global
+- The non-global operator is the only installation or `Subscription` for that operator type in the cluster
+
+#### Operator managed cleanup (not in scope for this proposal)
+
+There could be a use case where operator authors might require custom logic for deleting CRs e.g to order the deletion, or filter certain CRs from deletion to preserve state.
+
+For that `spec.operandCleanup` could be extended to another type, e.g `type: operator-managed`, which could allow a user to configure a job with custom logic for the deletion of CRs.
+The exact workflow and UX are unclear at this time, and are left as a future extension.
+
+## Design Details
+
+### Test Plan
+
+OLM's e2e testing suite should be upgraded to handle the following use cases:
+
+- An operator that is uninstalled with `spec.operandCleanup` unspecified or `none` preserves the CRs and CRD
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` removes all CRs and CRDs
+- An operator that is uninstalled with `spec.operandCleanup: olm-managed` and exceeds a timeout should generate an alert event, and not block operator removal
+
+### Risks and Mitigations
+
+It's worth considering if the clean up timeout `spec.operandCleanup.timeout` should have a "sane" upper bound to avoid having the operator uninstall be blocked by cleanup indefinitely.
+Or if users would require setting an inifinte timeout to block operator removal, if they want to inspect operator logs in the event of a cleanup failure.
```

Comment:
@mhrivnak That's a good point. You're right that timing out and removing the operator isn't helpful when the operator fails or is stalled on its own clean up.

The motivation for the timeout is not really a user story, and more from [one of the goals listed above](https://github.com/operator-framework/enhancements/pull/46/files/1a48736d607ecd81dc907b58920ac56133f9400a#diff-3824b2623f1ada24ff884541902a3edaR51):
```
After the operand removal timeout the operator should still be uninstalled
```
I'm not a 100% sure why @dmesser has this as a goal, but my guess is we don't want to break OLM's existing uninstall behavior with this clean up option, and block operator deletion indefinitely.

> In the general case, wouldn't it be better to show some condition or status, fire an alert, and allow someone or some thing to observe that state and decide on appropriate action?

I have the alert part outlined in a [story above](https://github.com/operator-framework/enhancements/pull/46/files/1a48736d607ecd81dc907b58920ac56133f9400a#diff-3824b2623f1ada24ff884541902a3edaR71).
```
As an operator user I want to see an alert event when OLM does not succeed in operand cleanup after a configured timeout.
```
Also [above](https://github.com/operator-framework/enhancements/pull/46/files/1a48736d607ecd81dc907b58920ac56133f9400a#diff-3824b2623f1ada24ff884541902a3edaR111):
```
If the cleanup timeout is exceeded, OLM should generate an alert event indicating a cleanup failure for that operator, and then proceed to uninstall the operator.
```
But it's more of a post-mortem at that point where the only action you can now take is to manually cleanup.

Although given this and the other problems that a timeout introduces, I'm not sure if preserving the existing uninstall UX is a great reason to do that.

So yeah this probably needs a bit more discussion to figure out if this really should be a goal of this enhancement.

Reply:
If the operator is unable to complete its removal of resources, there may be some simple reason that it can express and enable the user to resolve. It's not necessarily a post-mortem in that situation.

Reply:
maybe the doc can suggest an opt-in to write the respective [Condition CR](https://github.com/operator-framework/enhancements/blob/master/enhancements/operator-conditions.md) in this situation

Reply:
I see this was mentioned as a comment before: https://github.com/operator-framework/enhancements/pull/46/files#r483580878

Reply:
I think @mhrivnak convinced me that removing the Operator despite a timeout being been reached is not a good idea. I guess personally I'd still favor a system of telling me that it is not progressing any further after a sane timeout has been hit. It is common UX issue with declarative systems as the user cannot quite tell if it is just taking very long to reach desired state (and that is to be expected) or the system will hang indefinitely (which should be considered unexpected)

But in terms of alignment with standard Kube controller behavior we should probably follow what @mhrivnak proposed: hang indefinitely and let higher level systems or users decide how they want to progress. Such a higher level system (cluster installer) could then implement a timeout.

Reply:
Timeouts have been removed from the scope of this proposal per the latest changes.

Also added a line about having a special operator condition on the Operator Conditions CR that the operator can write to indicate when cleanup is stalled. But that's more of a future goal and not something that's directly implementable via this proposal.